src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -1335,6 +1335,53 @@ bool ONNXMaxPoolSingleOutOp::inferShapes() {
 
 //===----------------------------------------------------------------------===//
 
+bool ONNXPadOp::inferShapes() {
+  // Cannot infer shape if no shape exists.
+  if (!data().getType().isa<RankedTensorType>()) {
+    emitError("Pad: unknown input shape");
+    return false;
+  }
+
+  // Cannot infer if the pads is not constant
+  DenseElementsAttr padsAttributes =
+      getAttr("pads").dyn_cast_or_null<mlir::DenseElementsAttr>();
+
+  if (!padsAttributes) {
+    emitError("Pad: unknown pads");
+    return false;
+  }
+
+  auto dataTy = data().getType().cast<RankedTensorType>();
+  auto dataShape = dataTy.getShape();
+  auto dataRank = dataShape.size();
+  SmallVector<int64_t, 4> outputShape(dataShape.begin(), dataShape.end());
+
+  // Get pads from valueAttribute.
+  SmallVector<int64_t, 2> pads(dataRank * 2, -1);
+  auto valueIt = padsAttributes.getValues<IntegerAttr>().begin();
+  for (int i = 0; i < dataRank * 2; ++i)
+    pads[i] = (*valueIt++).cast<IntegerAttr>().getInt();
+
+  // Pads consists of two values for each axis of data.
+  // The two values specify the number of elements padded before and after
+  // respectively.
+  for (int i = 0; i < dataRank; ++i) {
+    int64_t p1 = pads[i];
+    int64_t p2 = pads[i + dataRank];
+    // Have to non-negative constant
+    if (p1 < 0 || p2 < 0) {
```

Comment:
I feel like the pad op lowering can be generic enough to handle the negative padding case. Is this too much effort?

Reply:
Fixed

Reply:
The current lower does not work with negative padding. When the affine condition is used, negative padding can be supported. Just keep the condition for now.